### PR TITLE
Extend yast2_software_management.pm

### PR DESCRIPTION
**Pull request by Zaoliang**
http://e13.suse.de/tests/2087 (sles 15)
http://e13.suse.de/tests/2092  (openSUSE TW)
see poo# 20966
needle PR:
https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/815
https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/361

### CODE REVIEW

File: tests/yast2_gui/yast2_software_management.pm [hunk 1]
```diff
@@ -9,16 +9,124 @@
...
+    assert_and_click 'sw_file_open';
+    assert_and_click 'sw_file_import';
+    assert_and_click 'package-load-package-list_cancel';
+    wait_still_screen 60;
+
+    # Package - All packages - Update
+    assert_and_click 'sw_package';
+    # use mouse_set and click All in the List because assert_and_click won't hold
+    # the menu open for update if newer version available
+    mouse_set(184, 354);
```
Comment (rwx788) [inline tests/yast2_gui/yast2_software_management.pm]: This won't scale, we should never use coordinates. What is exactly the problem here? Should not be too hard to get it in the os-autoinst

File: tests/yast2_gui/yast2_software_management.pm [hunk 2]
```diff
@@ -9,16 +9,124 @@
...
+    assert_and_click 'special-keys-overview-ok';
+
+    # View - Pattern
+    assert_and_click 'sw_view';
+    assert_and_click 'view_patterns';
+    assert_screen 'pattern-list';
+
+    # select a package to install: apache2
+    assert_and_click 'sw_search';
+    mouse_set(144, 183);
```
Comment (rwx788) [inline tests/yast2_gui/yast2_software_management.pm]: Same here

File: tests/yast2_gui/yast2_software_management.pm [hunk 3]
```diff
@@ -9,16 +9,124 @@
...
+    if (match_has_tag('automatic-changes')) {
+        send_key 'alt-o';
+        assert_screen [qw(unsupported-packages sw_management)];
+        if (match_has_tag('unsupported-packages')) {
+            send_key 'alt-o';
+        }
+        assert_screen 'installation-report', 180;
+        #Done and exit
+        send_key 'alt-f';
+        wait_still_screen 120;
```
Comment (rwx788) [inline tests/yast2_gui/yast2_software_management.pm]: Why we need this? Can it be replaced with screen assertion?

File: tests/yast2_gui/yast2_software_management.pm [hunk 4]
```diff
@@ -9,16 +9,124 @@
...
+    wait_still_screen;
+    send_key 'alt-f';
+    assert_screen [qw(automatic-changes sw_management)];
+    if (match_has_tag('automatic-changes')) {
+        send_key 'alt-o';
+        assert_screen [qw(unsupported-packages sw_management)];
+        if (match_has_tag('unsupported-packages')) {
+            send_key 'alt-o';
+        }
+        assert_screen 'installation-report', 180;
```
Comment (rwx788) [inline tests/yast2_gui/yast2_software_management.pm]: So we assert sw_management and if it matches we assert installation-report even if we didn't make any step. So should it be installation-report or they are same and then we need to assert again only if ```match_has_tag('unsupported-packages')``` is true.

File: tests/yast2_gui/yast2_software_management.pm (hunk 1 above)
Comment (Zaoliang) [reply tests/yast2_gui/yast2_software_management.pm]: assert_and_click won't keep menu open, so this is workaround. please let me know if you have a better solution for this.

File: tests/yast2_gui/yast2_software_management.pm (hunk 3 above)
Comment (Zaoliang) [reply tests/yast2_gui/yast2_software_management.pm]: this is somehow not working atm. I consider to change this and will try to check generic-desktop directly.

File: tests/yast2_gui/yast2_software_management.pm (hunk 4 above)
Comment (Zaoliang) [reply tests/yast2_gui/yast2_software_management.pm]: will change here as well: assert_screen  [qw(unsupport-packages installation-report)]

and take send_key 'alt-f' out of if {}

**Comment by okurz** [comment]
please update PR description

File: tests/yast2_gui/yast2_software_management.pm [hunk 5]
```diff
@@ -9,16 +9,129 @@
...
+    #Done and exit
+    assert_screen [qw(automatic-changes installation-report)];
+    if (match_has_tag('automatic-changes')) {
+        send_key 'alt-o';
+        assert_screen [qw(unsupported-packages installation-report)];
+        if (match_has_tag('unsupported-packages')) {
+            send_key 'alt-o';
+            wait_still_screen 120;
+            assert_and_click 'sw_finish';
+            wait_still_screen 60;
```
Comment (okurz) [inline tests/yast2_gui/yast2_software_management.pm]: wait_still_screen is a bad idea in gui tests and any stilltime with value 60 or 120 s does not make any sense as at least the clock timer on the desktop is changing once every 60 seconds so this call iteslf can never succeed so it is no better than a sleep 60s and giving a false impression

File: tests/yast2_gui/yast2_software_management.pm [hunk 6]
```diff
@@ -9,16 +9,129 @@
...
+    send_key 'tab';
+    send_key_until_needlematch 'update-if-newer-version-available', 'down';
+    send_key 'ret';
+    # make sure that package menu got closed after return
+    # otherwise following needle can be matched at wrong place
+    assert_screen [qw(all-in-this-list_still-selected sw_configurations)];
+    if (match_has_tag 'all-in-this-list_still-selected') {
+        mouse_click 'to-close-menu';
+    }
+    wait_still_screen;
```
Comment (okurz) [inline tests/yast2_gui/yast2_software_management.pm]: A wait_still_screen with default stilltime is IMHO just wasting time. Can we not wait for something specfic? E.g. a needle?

File: tests/yast2_gui/yast2_software_management.pm [hunk 7]
```diff
@@ -9,16 +9,129 @@
...
+        }
+    }
+    wait_still_screen 60;
+}
 
 sub run {
     my $self = shift;
     $self->launch_yast2_module_x11('sw_single', match_timeout => 120);
-    send_key "alt-a";    # Accept => Exit
+    if (is_sle('12-SP4+') || is_tumbleweed) {
```
Comment (okurz) [inline tests/yast2_gui/yast2_software_management.pm]: There should be a comment here why leap is not in

**Comment by okurz** [comment]
Nice idea but I am afraid the test could be very costly to maintain, is already taking multiple minutes and hardly checking any real system integration features but just clicking around in the GUI.

@rwx788 this might be a good example to discuss with the YaST team how useful they think this test could be. Could you check with them? We can take our time to cover this question so no hurry

**Comment by rwx788** [comment]
@okurz as we saw clicking actually find those NPE kind of issues, so there is a value in it, but it should be preferable covered by unit tests. So I would also keep it simple and verify actions we do, like enabling/disabling repos, installing packages.

**Comment by okurz** [comment]
yes, makes sense. But I think currently there are too many intermediate steps which make the needle maintenance hard and there is no checking of the result of the selected actions, e.g. that the package was installed or the repo was actually disabled

**Comment by Zaoliang** [comment]
@okurz updated now, thanks

**Comment by foursixnine** [comment]
I wonder if this is still relevant...

**Comment by rwx788** [comment]
I guess, but we decided not to enable it in prod, as it's a lot of maintenance. We could have enabled it in development job group, but I would wait for YaST UI framework we are working on.

**Comment by Zaoliang** [comment]
since this is not used and it was really a long time ago. I close this for now.

**Pull request closed without merging**